File: spx-backend/internal/model/record.go
package model

// Record is the model for project recording videos.
type Record struct {
	Model

	// UserID is the ID of the user who created the record.
	UserID int64 `gorm:"column:user_id;index;index:,composite:user_id_name,unique"`
	User   User  `gorm:"foreignKey:UserID"`

	// ProjectID is the ID of the project that the record is associated with.
	ProjectID int64   `gorm:"column:project_id;index"`
	Project   Project `gorm:"foreignKey:ProjectID"`

	// Name is the unique name of the record under the user.
	Name string `gorm:"column:name;index:,class:FULLTEXT;index:,composite:user_id_name,unique"`

	// Title is the display title of the record.
	Title string `gorm:"column:title"`

	// Description is the brief description of the record.
	Description string `gorm:"column:description"`

	// VideoURL is the URL of the recorded video file.
	VideoURL string `gorm:"column:video_url"`

	// ThumbnailURL is the URL of the thumbnail image.
	ThumbnailURL string `gorm:"column:thumbnail_url"`

	// Duration is the duration of the video in seconds.
	Duration int64 `gorm:"column:duration"`

	// FileSize is the size of the video file in bytes.
	FileSize int64 `gorm:"column:file_size"`

	// Visibility is the visibility of the record.
	Visibility Visibility `gorm:"column:visibility;index;default:1"` // default public

	// ViewCount is the number of times the record has been viewed.
	ViewCount int64 `gorm:"column:view_count;index"`

	// LikeCount is the number of times the record has been liked.
	LikeCount int64 `gorm:"column:like_count;index"`

	// Migration only fields. Do not use in application code.
	MO__deleted_at_is_null _deleted_at_is_null `gorm:"->:false;<-:false;column:_deleted_at_is_null;index:,composite:user_id_name,unique"`
}

// TableName implements [gorm.io/gorm/schema.Tabler].
func (Record) TableName() string {
	return "record"
}

File: spx-gui/src/apis/record.ts
import type { ByPage, PaginationParams } from './common'
import { client, ownerAll, Visibility } from './common'
import type { ProjectData } from './project'

export { Visibility, ownerAll }

export type RecordData = {
    /** Unique identifier */
    id: string
    /** Creation timestamp */
    createdAt: string
    /** Last update timestamp */
    updatedAt: string
    /** Unique username of the user who created the record */
    owner: string
    /** ID of the associated project */
    projectId: number
    /** Associated project data (optional, included in some contexts) */
    project?: ProjectData
    /** Unique name of the record under the user */
    name: string
    /** Display title of the record */
    title: string
    /** Brief description of the record */
    description: string
    /** URL of the recorded video file */
    videoUrl: string
    /** URL of the thumbnail image */
    thumbnailUrl: string
    /** Duration of the video in seconds */
    duration: number
    /** Size of the video file in bytes */
    fileSize: number
    /** Visibility of the record */
    visibility: Visibility
    /** Number of times the record has been viewed */
    viewCount: number
    /** Number of likes the record has received */
    likeCount: number
}

export type CreateRecordParams = {
    /** ID of the project that the record is associated with */
    projectId: number
    /** Unique name of the record under the user */
    name: string
    /** Display title of the record */
    title: string
    /** Brief description of the record */
    description: string
    /** URL of the recorded video file */
    videoUrl: string
    /** URL of the thumbnail image */
    thumbnailUrl: string
    /** Duration of the video in seconds */
    duration: number
    /** Size of the video file in bytes */
    fileSize: number
}

export async function createRecord(params: CreateRecordParams, signal?: AbortSignal) {
    return client.post('/records', params, { signal }) as Promise<RecordData>
}

export type UpdateRecordParams = Partial<Pick<RecordData, 'title' | 'description' | 'visibility'>>

export async function updateRecord(owner: string, name: string, params: UpdateRecordParams, signal?: AbortSignal) {
    return client.put(`/record/${encodeURIComponent(owner)}/${encodeURIComponent(name)}`, params, {
        signal
    }) as Promise<RecordData>
}

export function deleteRecord(owner: string, name: string) {
    return client.delete(`/record/${encodeURIComponent(owner)}/${encodeURIComponent(name)}`) as Promise<void>
}

export type ListRecordParams = PaginationParams & {
    /**
     * Filter records by the owner's username.
     * Defaults to the authenticated user if not specified. Use * to include records from all users.
     **/
    owner?: string
    /** Filter records by associated project (format: owner/project) */
    projectFullName?: string
    /** Filter records by name pattern */
    keyword?: string
    /** Field by which to order the results */
    orderBy?: 'createdAt' | 'updatedAt' | 'duration' | 'viewCount' | 'likeCount' | 'likedAt'
    /** Order in which to sort the results */
    sortOrder?: 'asc' | 'desc'
    /** Filter records liked by the specified user */
    liker?: string
}

export async function listRecord(params?: ListRecordParams) {
    return client.get('/records/list', params) as Promise<ByPage<RecordData>>
}

export async function getRecord(owner: string, name: string, signal?: AbortSignal) {
    return client.get(`/record/${encodeURIComponent(owner)}/${encodeURIComponent(name)}`, undefined, {
        signal
    }) as Promise<RecordData>
}
// export async function getRecord(owner: string, name: string): Promise<RecordData> {
//     return client.get(`/records/${encodeURIComponent(owner)}/${encodeURIComponent(name)}`) as Promise<RecordData>
// }

/** Record a view for the given record */
export async function recordRecordView(owner: string, name: string) {
    return client.post(`/record/${encodeURIComponent(owner)}/${encodeURIComponent(name)}/view`) as Promise<void>
}

/**
 * Check if given record liked by current logged-in user.
 * If not logged in, `false` will be returned.
 */
export async function isLikingRecord(owner: string, name: string) {
    try {
        await client.get(`/record/${encodeURIComponent(owner)}/${encodeURIComponent(name)}/liking`)
        return true
    } catch (e) {
        // Handle API exceptions similar to project.ts
        // Not liked or not logged in
        return false
    }
}

/** Like given record as current logged-in user */
export async function likeRecord(owner: string, name: string) {
    return client.post(`/record/${encodeURIComponent(owner)}/${encodeURIComponent(name)}/liking`) as Promise<void>
}

/** Unlike given record as current logged-in user */
export async function unlikeRecord(owner: string, name: string) {
    return client.delete(`/record/${encodeURIComponent(owner)}/${encodeURIComponent(name)}/liking`) as Promise<void>
}

export function parseRecordFullName(fullName: string) {
    const [encodedOwner, encodedName] = fullName.split('/')
    const owner = decodeURIComponent(encodedOwner)
    const name = decodeURIComponent(encodedName)
    return { owner, name }
}

export function stringifyRecordFullName(owner: string, name: string) {
    const encodedOwner = encodeURIComponent(owner)
    const encodedName = encodeURIComponent(name)
    return `${encodedOwner}/${encodedName}`
}